### DISCUSSION TITLE
static, non-static의 차이

### DISCUSSION THREAD

**Discussion by GIVEN53** [2024-woowacourse-study/level-interview#19] [Java]
.

**Comment by nayonsoso** [comment]
크게 세가지 차이점이 있습니다. 
첫번째는, '생성되는 시점이 다르다'는 것이고 
두번째는, '클래스와 객체 중 어디에 속하는지'입니다. 
마지막으로 'JVM 안에서 저장되는 공간이 다르다'는 점도 있습니다.

첫째로, static은 JVM이 클래스를 로드하는 시점에 생성되고, 프로그램이 종료될 때까지 유지됩니다.
non-static은 객체가 생성될 때 생성되고, GC 에 의해 관리됩니다.
이런 특성 때문에 static 메서드에서는 this 키워드를 쓸 수 없습니다.
this는 지금 작동하는 객체를 의미하는데, static 함수가 로드되는 시점에서는 this가 있을 수 없으니깐요.

둘째로, staic은 클래스 레벨에 속합니다. 때문에 모든 객체가 같은 값을 공유합니다.
이런 특성 떄문에 인스턴스를 생성하지 않고 클래스를 통해 바로 접근할 수 있습니다.
반면, non-static은 객체 레벨에 속합니다. 객체마다 다른 값을 가질 수 있습니다.

셋째로, static은 메소드 영역에 속합니다. 
이 영역은 GC에 의해 관리되지 않는 공간이기 때문에 static을 남용하면 메모리 부족 문제가 생길 수 있습니다.
반면 non-static은 Heap에 저장되기 때문에 GC의 관리를 받습니다.

**Comment by jimi567** [comment]
static 멤버(메소드), non-static 멤버(메소드)의 차이는 다음과 같다.

static은 클래스에 속하고, non-static은 객체에 속하게 된다.
초기화 시점은 static의 경우 클래스가 로딩될 때 초기화되며, 프로그램 종료까지 메모리에 유지된다.
반대로 non-static은 객체가 생성될 때, 초기화 되며, 객체가 삭제될 때까지 메모리에 유지된다.

static은 non-static과 달리 클래스가 로딩할 때 생성되므로, 클래스의 인스턴스(객체)를 생성하지 않고도 사용할 수 있다.

용도는 static의 경우 주로 유틸리티 메서드 등을 표현하거나, 상수 값을 저장하는데 사용된다. 또, 객체를 생성하기 전에 사용해야하는 작업을 수행할 때 사용하게 된다. 대표적인 예로 정적 팩토리 메소드가 있다.
non-static은 객체의 상태를 표현하거나, 객체의 상태에 따라 동작하는 작업을 수행할 때 사용한다.

**Comment by kunsanglee** [comment]
### static (정적 멤버)
정적 멤버는 클래스가 메모리에 로드될 때 한 번만 생성되고 초기화됩니다. 이들은 메서드 영역(Method Area)에 저장됩니다. 정적 변수나 메서드는 클래스의 모든 인스턴스에 의해 공유되며, 이들은 클래스 단위로 존재하기 때문에 인스턴스 간에 데이터를 공유하는 데 사용될 수 있습니다.

- 정적 변수: 모든 인스턴스에 의해 공유되는 변수입니다. 클래스가 메모리에 로드될 때 정적 변수가 메서드 영역에 할당되며, 프로그램이 종료될 때까지 존재합니다. 이는 클래스의 상태를 모든 인스턴스가 공유한다는 것을 의미합니다.
- 정적 메서드: 클래스의 인스턴스를 생성하지 않고도 호출할 수 있는 메서드입니다. this 참조가 없기 때문에 인스턴스 변수나 메서드에 접근할 수 없습니다. 정적 메서드는 주로 유틸리티 함수나 인스턴스와 무관한 작업을 수행하는 데 사용됩니다. 정적 메서드의 코드도 메서드 영역에 저장되며, 클래스에 속한 모든 인스턴스가 이 코드를 공유합니다.

### non-static (비정적 멤버)
비정적 멤버 (인스턴스 멤버)는 객체가 생성될 때마다 각 인스턴스에 대해 별도로 메모리가 할당됩니다. 이런 멤버들은 힙 메모리에 저장됩니다. 각각의 객체는 자신만의 인스턴스 변수 복사본을 가지고 있으며, 이 변수들은 객체가 생성될 때 힙 메모리에 할당됩니다.

- 인스턴스 변수: 각 객체의 고유한 상태를 저장하는 변수입니다. 이 변수들은 객체가 생성될 때 힙 메모리에 할당되며, 객체가 더 이상 참조되지 않고 가비지 컬렉터에 의해 수집될 때까지 존재합니다.
- 인스턴스 메서드: 객체의 인스턴스 변수에 접근하고 조작할 수 있는 메서드입니다. 이 메서드는 객체의 상태와 행동을 정의합니다. 메서드의 코드 자체는 모든 인스턴스에 의해 공유되지만, 'this' 참조를 통해 호출된 객체의 인스턴스 변수에 접근하여, 호출된 객체의 상태에 따라 다르게 작동합니다. 인스턴스 메서드의 코드 역시 메서드 영역에 저장되지만, 실행 시 'this'를 통해 특정 인스턴스에 속한 상태와 연동됩니다.

**Comment by GIVEN53** [comment]
static
클래스 로드
메서드 영역
모든 객체 공유

non-static
객체 생성
힙 영역 - GC 대상
객체 고유
